src/Microsoft.DotNet.Build.Tasks/ValidateProjectDependencyVersions.cs

```diff
@@ -0,0 +1,127 @@
+
+            public string FindValidationError(string packageId, string version)
+            {
+                var dependencyVersionRange = VersionRange.Parse(version);
+                NuGetVersion dependencyVersion = dependencyVersionRange.MinVersion;
+
+                if (_idPattern.IsMatch(packageId))
+                {
+                    if (!string.IsNullOrWhiteSpace(_expectedVersion) && _expectedVersion != version)
+                    {
+                        return string.Format(
+                            "package version '{0}', but expected '{1}' for packages matching '{2}'",
+                            version,
+                            _expectedVersion,
+                            _idPattern);
+                    }
+                    if (!string.IsNullOrWhiteSpace(_expectedPrerelease) &&
+                        dependencyVersion.IsPrerelease &&
+                        _expectedPrerelease != dependencyVersion.Release)
+                    {
+                        return string.Format(
+                            "package prerelease '{0}', but expected '{1}' for packages matching '{2}'",
+                            dependencyVersion.Release,
+                            _expectedPrerelease,
+                            _idPattern);
+                    }
+                }
+                return null;
+            }
+        }
+
+        [Required]
+        public ITaskItem ProjectJson { get; set; }
+
+        public bool ProhibitFloatingDependencies { get; set; }
+
+        public ITaskItem[] ValidationPatterns { get; set; }
+
+        public override bool Execute()
+        {
+            var patterns = ValidationPatterns.Select(item => new ValidationPattern(item));
+
+            using (TextReader projectFileReader = File.OpenText(ProjectJson.ItemSpec))
+            {
+                var projectJsonReader = new JsonTextReader(projectFileReader);
+
+                var serializer = new JsonSerializer();
+                JObject projectRoot = serializer.Deserialize<JObject>(projectJsonReader);
+
+                foreach (var package in projectRoot["dependencies"].OfType<JProperty>())
+                {
+                    string id = package.Name;
+                    string version = package.Value.ToObject<string>();
+
+                    string versionMessage = string.Format(
+                        "{0} {1} in {2}",
+                        id,
+                        version,
+                        ProjectJson);
+
+                    if (ProhibitFloatingDependencies && version.Contains('*'))
+                    {
+                        Log.LogError("Floating dependency detected: {0}", versionMessage);
+                    }
+                    else
+                    {
+                        string validationError = patterns
+                            .Select(pattern => pattern.FindValidationError(id, version))
+                            .FirstOrDefault(error => error != null);
```

Comment:
You should iterate through all of these and log all of them.  Only logging the first one is a recipe for "wack-a-mole" workflow.